src/Pos/Txp/Worker.hs

```diff
@@ -0,0 +1,126 @@
+{-# LANGUAGE CPP                 #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+
+module Pos.Txp.Worker
+       ( txpWorkers
+       ) where
+
+import           Universum
+
+import           Data.IORef        (newIORef, readIORef, writeIORef)
+import           Data.Time.Units   (Second, convertUnit)
+import           Mockable          (delay, throw)
+import           System.Wlog       (logWarning)
+
+import           Pos.Communication (ConversationActions (..), DataMsg (..), InvMsg (..),
+                                    InvOrData, Limit, LimitType, LimitedLengthExt,
+                                    MempoolMsg, OutSpecs,
+                                    RelayError (UnexpectedData, UnexpectedInv),
+                                    RelayProxy (..), ReqMsg (..), SendActions,
+                                    TxMsgContents, TxMsgTag, WorkerSpec, convH,
+                                    handleDataL, handleInvL, reifyMsgLimit, toOutSpecs,
+                                    withLimitedLength, worker)
+import           Pos.Constants     (isDevelopment)
+import           Pos.DHT           (DHTNode, converseToNode, getKnownPeers)
+import           Pos.Slotting      (getLastKnownSlotDuration)
+import           Pos.Ssc.Class     (SscWorkersClass)
+import           Pos.Txp.Core      (TxId)
+import           Pos.Util          (mconcatPair)
+import           Pos.WorkMode      (WorkMode)
+
+-- | All workers specific to transaction processing.
+txpWorkers
+    :: (SscWorkersClass ssc, WorkMode ssc m)
+    => ([WorkerSpec m], OutSpecs)
+txpWorkers =
+    merge $ []
+#if defined(WITH_WALLET)
+            ++ [ queryTxsWorker | not isDevelopment ]
+#endif
+  where
+    merge = mconcatPair . map (first pure)
+
+#if defined(WITH_WALLET)
+
+-- | When we're behind NAT, other nodes can't send data to us and thus we
+-- won't get transactions that are broadcast through the network – we have to
+-- reach out to other nodes by ourselves.
+--
+-- This worker just triggers every @max (slotDur / 4) 5@ seconds and asks for
+-- tx mempool.
+queryTxsWorker
+    :: (WorkMode ssc m, SscWorkersClass ssc)
+    => (WorkerSpec m, OutSpecs)
+queryTxsWorker = worker requestTxsOuts $ \sendActions -> do
+    slotDur <- getLastKnownSlotDuration
+    nodesRef <- liftIO . newIORef =<< getKnownPeers
+    let delayInterval = max (slotDur `div` 4) (convertUnit (5 :: Second))
+        action = forever $ do
+            -- If we ran out of nodes to query, refresh the list
+            whenM (null <$> liftIO (readIORef nodesRef)) $
+                liftIO . writeIORef nodesRef =<< getKnownPeers
+            -- If we managed to get any nodes (or if the list wasn't empty in
+            -- the first place), we ask the first node for the tx mempool.
+            liftIO (readIORef nodesRef) >>= \case
+                []          -> return ()
+                (node:rest) -> do
```

Comment:
As @Martoon-00 said above, if the node closes the connection this code won't hang. However, if it accepts and sends nothing, then, yes, it will hang. I don't know how to add a timeout, but hopefully @Martoon-00 will add timeouts to all connections and I won't have to do anything.